GPIO Driver

john-zapanta-jr opened the pull request:
# Description

GPIO Driver, example configs, documentation, and pytest added.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit test using mock pins added.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

SKalt committed:
fix(tests): got ./ci-integration/virtualization/Dockerfile to build

Notes from investigating `The command 'bash -c pip install -e . --user' returned a non-zero code: 1`:
./ci-integration/virtualization/Dockerfile was running off of debian 10, aka buster.  Buster is possibly the last debian to have python as python2.7 rather than 3.x.  Thus, `pip` runs as python 2.7 and fails to work with ./setup.py, which is written to work in python 3.x.  That raised a new error:
```
Exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/_internal/req/req_install.py", line 339, in check_if_exists
    self.satisfied_by = pkg_resources.get_distribution(str(no_marker))
  File "/usr/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 481, in get_distribution
    dist = get_provider(dist)
  File "/usr/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 357, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/usr/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (wheel 0.32.3 (/usr/lib/python3/dist-packages), Requirement.parse('wheel==0.30'), {'volttron'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/base_command.py", line 143, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/_internal/commands/install.py", line 338, in run
    resolver.resolve(requirement_set)
  File "/usr/lib/python3/dist-packages/pip/_internal/resolve.py", line 102, in resolve
    self._resolve_one(requirement_set, req)
  File "/usr/lib/python3/dist-packages/pip/_internal/resolve.py", line 256, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "/usr/lib/python3/dist-packages/pip/_internal/resolve.py", line 193, in _get_abstract_dist_for
    req, self.require_hashes, self.use_user_site, self.finder,
  File "/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py", line 329, in prepare_editable_requirement
    req.check_if_exists(use_user_site)
  File "/usr/lib/python3/dist-packages/pip/_internal/req/req_install.py", line 350, in check_if_exists
    self.req.name
  File "/usr/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 481, in get_distribution
    dist = get_provider(dist)
  File "/usr/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 357, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/usr/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (wheel 0.32.3 (/usr/lib/python3/dist-packages), Requirement.parse('wheel==0.30'), {'volttron'})
```
I fixed this by loosening the version requirements on the `wheel` package.

SKalt committed:
fix(tests):  configure CI to report errors

I've also fixed some of the warnings that `shellcheck` reports.

john-zapanta-jr committed:
GPIO Driver developed, documentation added, test added, and example configurations added.

John Zapanta Jr.

craig8 committed:
Added prototype docker wrapper and fixture for docker

craigpnnl committed:
keep wheel @ 0.30

jklarson commented on services/core/MasterDriverAgent/master_driver/interfaces/gpio.py:
```diff
@@ -0,0 +1,542 @@
+        self.config_dict = None
+        self.driver_path = None
+        # GPIO registers are of non-standard datatypes, so override existing register type dictionary
+        self.registers = {
```
This may not be required ( and in fact may be something I need to change for Ecobee). In Ecobee this was done for register types that returned data types that didn't match the existing register types (for example: program and vacation endpoints return dictionary structures - that said the holds and settings from Ecobee should probably be int registers).

jklarson commented on services/core/MasterDriverAgent/master_driver/interfaces/gpio.py:
```diff
@@ -0,0 +1,542 @@
+        :param config_dict: Dictionary containing configuration information.
+        :param registry_config_str: List of dictionaries containing point information used to establish registers.
+        """
+        self.config_dict = config_dict
```
I received feedback from Craig that we should be using self.config.update(config_dict) here on the updates I'm doing for the Ecobee driver.

jklarson commented on services/core/MasterDriverAgent/master_driver/interfaces/gpio.py:
```diff
@@ -0,0 +1,542 @@
+                )
+                continue
+            if int(reg["Pin"]) < 2 or int(reg["Pin"]) > 27:
+                raise ValueError(
```
I think we could merge this conditional in the try and capture both states with the error message. Ideally I think we'd have something describing valid pin numbers: "Registry config entry {} provided an invalid pin. GPIO supports Broadcom pins as integers from 2 to 27" or something to that effect.

jklarson commented on services/core/MasterDriverAgent/master_driver/interfaces/gpio.py:
```diff
@@ -0,0 +1,542 @@
+
+            # Default setting for device type is GPIO, which is the broadest category supported.
+            if not reg.get("Device Type"):
+                reg["Device Type"] = "output"
```
Lets use something like string.lower() to try to capture more of the potential input by a user (e.i. "Output" would fail here, but could easily be checked for with .lower() )

john-zapanta-jr replied:
It's actually not a conditional here, it's an assignment to the value "output" by default.

jklarson commented on services/core/MasterDriverAgent/master_driver/interfaces/gpio.py:
```diff
@@ -0,0 +1,542 @@
+                                          "float is required").format(index))
+
+                # Validate pull up and active high config.
+                if not (reg.get("Pull Up") is None):
```
This looks like it should be reg.get("Pull Up") is not None:

jklarson commented on services/core/MasterDriverAgent/master_driver/interfaces/gpio.py:
```diff
@@ -0,0 +1,542 @@
+                                                          description=description))
+
+            # Insert digital and generic output GPIO registers.
+            elif reg["Device Type"] == "output" or reg["Device Type"] == "digital_output":
```
Lets have another .lower() style comparison here.

jklarson commented on services/core/MasterDriverAgent/master_driver/interfaces/gpio.py:
```diff
@@ -0,0 +1,542 @@
+        if not register:
+            raise ValueError("Valid point name not provided.")
+        # Invoke the action method of the register if 'value' is a dictionary.
+        if isinstance(value, dict):
```
should this work with read-only points?

john-zapanta-jr replied:
Input points are read only, but have actions that need to be accessed through set point - i.e. call backs. I don't let values get set for read only points though.

jklarson commented on services/core/MasterDriverAgent/master_driver/interfaces/gpio.py:
```diff
@@ -0,0 +1,542 @@
+
+    def push_cov_to_master(self):
+        """
+        Helper function for performing a callback that publishes to the message bus.
```
I think here we can use self.driver.publish_cov_value(self.device.value, self.point_name)

jklarson commented on docs/source/core_services/drivers/driver_configuration/gpio_driver.rst:
```diff
@@ -0,0 +1,261 @@
+
+    Point Name - Name of a point as a user would like it to be referenced and displayed by Volttron.
+
+    Pin - The broadcom pin number of the GPIO pin the device is connected to.
```
Broadcom should be capitalized

jklarson commented on docs/source/core_services/drivers/driver_configuration/gpio_driver.rst:
```diff
@@ -0,0 +1,261 @@
+
+The following is an example publish:
+
+::
```
Try capturing output as displayed from the listener agent.

jklarson commented on services/core/MasterDriverAgent/tests/test_gpio_driver.py:
```diff
@@ -0,0 +1,243 @@
+    }
+]
+
+def drive_low(abort_pin):
```
We should have tests which also test failure conditions. What happens if we provide illegal values, etc. Pytest.raises and other tools can be handy for this.

Commit:
Requested changes addressed. Added input accommodation, adjusted driver callback, addressed typo, and added failure testing.

John Zapanta Jr.

SKalt committed:
fix: Revert wheel package to 0.30

Wheel >= 0.31 appears to include breaking changes.  As requested in PR#2363, I've reverted to 0.30.  Special thanks to @craig8, from whom I copied most of these changes.

SKalt committed:
Merge branch 'wip--make-sense-of-tests' of github.com:SKalt/volttron into wip--make-sense-of-tests

SKalt committed:
chore(tests): comment ultra-verbose bash logging

craigpnnl committed:
Merge pull request #2363 from SKalt/wip--make-sense-of-tests

Get CI runs to report failures

craig8 committed:
Modified rabbit_dependencies to get pinned version of erlang for debian based dists.

craig8 committed:
Merge branch 'develop' of https://github.com/VOLTTRON/volttron into bugfix/2374_pin_erlang

craigpnnl committed:
Update bootstrap.py

Pin to 0.30 wheel rather than ~=

Commit:
erlang version specified for raspbian differs from x86 distros

Commit:
fixed readme section heading

craigpnnl committed:
Merge pull request #45 from laroque/bugfix/erl-pin

erlang version specified for raspbian differs from x86 distros

Commit:
Added Stand Alone version of the FileWatcher.

Commit:
Added Stand Alone version of the FileWatcher.

Commit:
Modified volttron utils for file watcher to work successfully

Commit:
Modified volttron utils for file watcher to work successfully

Commit:
Updated Stand Alone File Watcher settings and code along with better comments

Commit:
Updated Stand Alone File Watcher settings and code along with better comments

Commit:
Minor grammatical fixes

Commit:
Minor grammatical fixes

Commit:
Sync fork.

-John Zapanta Jr.

craigpnnl committed:
Merge pull request #2375 from craig8/bugfix/2374_pin_erlang

Bugfix/2374 pin erlang

craigpnnl commented on docs/source/core_services/drivers/driver_configuration/gpio_driver.rst:
```diff
@@ -0,0 +1,270 @@
+GPIO Driver
+===========
+
+The GPIO driver is an implementation of a VOLTTRON driver frame work Interface. An instance of a VOLTTRON Interface
```
framework not frame work

craigpnnl committed:
Update docker_wrapper.py

craigpnnl committed:
Merge pull request #2366 from craig8/docker-fixture

Added prototype docker wrapper and fixture for docker

craigpnnl committed:
base requirements for rmq to work properly

craigpnnl committed:
Merge pull request #2382 from VOLTTRON/pin-gevent-requests-1

base requirements for rmq to work properly

shwethanidd committed:
Fixes issue#2345. AuthService attempts to get peerlist to send the auth updates to all peers. The peerlist is able to get the list of response from router, but setting asyn result object is not working properly. As workaround peerlist subsystem keeps track of list of peers to make it accessible to auth service

shwethanidd committed:
Update auth.py

Reduced sleep period

shwethanidd committed:
Update peerlist.py

Removed old comment

jklarson commented on services/core/MasterDriverAgent/tests/test_gpio_driver.py:
```diff
@@ -0,0 +1,309 @@
+def test_configuration():
+    test = Interface()
+    # Test configuration input validation.
+    with pytest.raises(ValueError):
```
Lets break out the failure cases into their own test and use pytest.mark.parametrize to improve code-reuse.

craigpnnl committed:
Merge remote-tracking branch 'sgilbride/stand_alone_csv' into stand_alone_csv

craigpnnl committed:
Merge pull request #2377 from sgilbride/stand_alone_csv

Stand Alone File Watch Publisher Agent

Commit:
Configuration tests split into passing and failing tests, and parametrization implemented for testing configuration variation.

John Zapanta Jr.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Syncing fork with upstream.

Commit:
Fixed typo in the GPIO driver documentation.

John Zapanta Jr.
